### DISCUSSION TITLE
Weird behavior in SyN registration

### DISCUSSION THREAD

**Discussion by m-albert** [dipy/dipy#1838] [Registration]
First of all thanks a lot for the great open source tools, I’m using dipy’s image registration extensively.

I have an issue / question regarding using SymmetricDiffeomorphicRegistration in combination with CCMetric.

I’m optimising the registration performance for my application (zebrafish embryos imaged with fluorescence microscopy). More specifically, I am playing around with the scaling factors of the scale space and the update field smoothing (parameter sigma_diff in CCMetric). To do so, I modified the registration process in such a way that I can manually set the current scaling factor and sigma_diff (taking advantage of the fact that dipy is written in accessible python!).

I found parameter sets which achieve significantly better performance for my dataset than the presets (meaning doubling of scaling factors in each level and constant sigma_diff). So far so good.

However, there’s an odd behaviour I don't understand. **Often when I decrease sigma_diff from one level to the next, the metric increases instead of decreasing.** However, when I run the two levels with different sigma_diffs in separate registration calls, the metric nicely decreases as expected and registration performance improves. Why is this and is this supposed to happen?

I tried to come up with a minimal example shown in this jupiter notebook: https://github.com/quakberto/dipy_syn/blob/master/test_syn_sigma_diffs.ipynb. The notebook also shows the custom code I use for creating the modified scale space.

Here’s a short summary:

I generate a random static image S and rotate it to get a moving image M
I run registrations with different sigma_diffs:
a) one level, constant sigma_diff = 1 (S,M)->M_a
b) one level, constant sigma_diff = 2 (S,M)->M_b
c) two levels, decreasing sigma_diffs = [2,1] (S,M)->M_c
d) one level, sigma_diff = 1, starting with M_b: (S,M_b)->M_c

Now, using the normalised cross correlation and also the sum of norms of differences between the experimental and ground truth displacement fields as metrics, surprisingly case c) performs worse than b). However, case d) greatly improves on the outcome of b). Here's a figure:

![image](https://user-images.githubusercontent.com/12528388/58029767-8d83bd80-7b1d-11e9-8706-68031afbfb89.png)

My question is: Why does the last level in case c) deteriorate the registration performance? And why does a fresh run behave so differently?

**Comment by skoudoro** [comment]
Hi @quakberto,

Thank you for this complete report and your IPython notebook example. Really nice feedback! 
After a really quick look at the `CCMetric`, I think I found a small error [here](https://github.com/nipy/dipy/blob/master/dipy/align/metrics.py#L315) (`energy` variable not updated for backward step, it should be `self.energy`). This might be the cause but we clearly need to look deeper into it and play with your example before giving you an answer.

**Comment by m-albert** [comment]
Thanks a lot for having a look @skoudoro :)

True, after the backward step the energy attribute is not updated. However, I was checking whether this would have an influence on the registration and realised that the energy is the same for the forward and backward steps. This is because it actually refers to the normalised cross correlation of the current images in the reference space (meaning before warping). So probably not setting the class attribute in the line you linked is even intentional.

Any other ideas? And in general, do you agree what I describe above is a weird behavior? Thanks again!
